benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/README.md

```diff
@@ -1,56 +1,76 @@
+the Kokkos source code, this can be set inside the makefile or with the
+`KOKKOS_PATH` environment variable. If you are using Bash and the Gnu
+compiler then:
 
-    > . ../../../../../compiler_setup/gnu.sh
+    > source ../../../../../compiler_setup/gnu.sh
     > export KOKKOS_PATH=<path to kokkos>
     > make
 
-should build the `nemolite2d_kokkos.exe` binary.
+will compile the default `nemolite2d_views_kokkos_OpenMP.exe` binary.
+
+Kokkos supports multiple execution spaces (e.g Serial, Threads, OpenMP,
+Cuda, HPX), these can be set for each Makefile target with the
+KOKKOS_DEVICES environment variable (examples below). If none is provided,
+it defaults to the OpenMP backend.
+
+Kokkos also has several abstractions to control where (execution spaces)
+and how (execution policies and execution patterns) computation is executed
+and how memory is laid out (Views). This make it possible to create very
+different Kokkos implementations, at the moment there are 2 versions:
+
+- Rawpointers version: This version uses the Kokkos parallel dispatch
+(execution policies and execution pattern) but in top of a raw pointer
+arrays (given by the dl_esm_inf infrastructure). This
+version is memory space efficient as it doesn't need to do extra copies
+between the Fortran and Kokkos parts of the code, but it is not able to
+run on the GPU or change the data layout. This version is available in
+`time_step_rawpointers_kokkos.cpp` and can be build with the command:
+
+    > make nemolite2d_rawpointers_kokkos
+
+- View Containers: This version uses the Kokkos View containers in addition
+to the Kokkos parallel dispatch. This allows Kokkos to control the data layout,
+the padding, and the synchonization between host and device (GPU execution)
+but it requieres additional copies of the data for each layer.
+This version is available in `time_step_views_kokkos.cpp` and can be built
+with an OpenMP or a Cuda backend by setting the KOKKOS_DEVICES environment
+variable:
+
+    > make nemolite2d_views_kokkos KOKKOS_DEVICES=OpenMP
+
+    > make nemolite2d_views_kokkos KOKKOS_DEVICES=Cuda
 
 ## Running ##
 
 Model parameters (size of domain [jpiglo,jpjglo], number of time-steps
 [nitend], whether or not and how often to do output [irecord]) may be
 configured by editing the `namelist` file.
 
-Kokkos supports multiple execution spaces (e.g Serial, Threads, OpenMP,
-Cuda, HPX) but the current implementation is limited to only OpenMP.
 
-Therefore, we can configure the parallel environment using the OpenMP
-environment variables. For example, the number of threads to use can be
-set via the OMP_NUM_THREADS environment variable.
+In the case of OpenMP we can configure the parallel environment using the
+standard OpenMP environment variables. For example, the number of threads
+to use can be set via the OMP_NUM_THREADS environment variable.
 
     > export OMP_NUM_THREADS=4
-    > ./nemolite2d_kokkos.exe
+    > ./nemolite2d_views_kokkos_OpenMP.exe
 
```

Comment:
Are there any settings required when running on a GPU?

Reply:
Not so far, although my current GPU executions are quite simple. I added a general instruction to run as `./nemolite2d_<version>_<device>.exe` the binaries produced by the Makefile.